File: library/src/main/java/aiven/guide/view/layer/Layer.java
package aiven.guide.view.layer;


import android.app.Activity;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;

import aiven.guide.view.clip.BaseClipPosition;
import aiven.guide.view.face.IntroPanel;
/**
 * Author: ChenYouSheng
 * Date: 2021/8/5
 * Email: chenyousheng@lizhi.fm
 * Desc: 引导层,处理裁剪区和指示区
 */
class Layer {
    @NonNull
    private ClipHolder clipHolder;
    @NonNull
    private FaceHolder faceHolder;
    @NonNull
    protected String tag;

    public Layer(String tag) {
        this.clipHolder = new ClipHolder();
        this.faceHolder = new FaceHolder();
        this.tag = tag;
    }


    public void setClipTarget(BaseClipPosition target){
        clipHolder.setTarget(target);
    }

    public void setFacePanel(IntroPanel facePanel){
        this.faceHolder.setFacePanel(facePanel);
    }

    /**
     * 图层绘制入口
     * @param canvas
     * @param paint
     * @param clip
     * @param parentWidth
     * @param parentHeight
     */
    public void draw(Canvas canvas, Paint paint, boolean clip, float parentWidth, float parentHeight){
        if(clip){
            clipHolder.draw(canvas,paint,null,parentWidth,parentHeight);
        }else{
            faceHolder.draw(canvas,paint, clipHolder.getRectF(),parentWidth,parentHeight);
        }
    }

    protected void build(@Nullable Activity activity) {
        if(clipHolder != null){
            clipHolder.build(activity);
        }
        if(faceHolder != null){
            faceHolder.build(activity);
        }
    }


    protected void build(@Nullable Fragment fragment) {
        if(clipHolder != null){
            clipHolder.build(fragment);
        }
        if(faceHolder != null){
            faceHolder.build(fragment);
        }
    }

    @NonNull
    public String getTag() {
        return tag;
    }

    public boolean isClipEventPassThrough(){
        if(clipHolder != null){
            return clipHolder.isEventPassThrough();
        }
        return false;
    }


    public boolean isTouchInClip(float x,float y){
        if(clipHolder != null && clipHolder.getRectF() != null){
            if(clipHolder.getRectF().contains(x,y)){
                return true;
            }
        }
        return false;
    }

    public boolean isTouchInIntro(float x,float y){
        if(faceHolder != null && faceHolder.getRectF() != null){
            if(faceHolder.getRectF().contains(x,y)){
                return true;
            }
        }
        return false;
    }
}

File: guide/src/main/java/com/example/guide/layer/Layer.kt
package com.example.guide.layer

import android.app.Activity
import android.graphics.Canvas
import android.graphics.Paint
import android.support.v4.app.Fragment
import com.example.guide.option.BaseClipOption
import com.example.guide.option.DescOption

/**
 * Author: ChenYouSheng
 * Date: 2021/8/5
 * Email: chenyousheng@lizhi.fm
 * Desc:  引导层,处理裁剪区和指示区
 */
class Layer(val tag: String) {
    private var clipHolder: ClipHolder = ClipHolder()
    private var descHolder: DescHolder = DescHolder()

    /**
     * 设置裁剪选项
     */
    fun setClipOption(option: BaseClipOption) {
        clipHolder.setTargetOption(option)
    }

    /**
     * 设置描述选项
     */
    fun setDescOption(option: DescOption) {
        descHolder.setTargetOption(option)
    }

    /**
     * 构建裁剪区域
     */
    fun build(activity: Activity?) {
        clipHolder.build(activity)
    }

    /**
     * 构建裁剪区域
     */
    fun build(fragment: Fragment?) {
        clipHolder.build(fragment)
    }

    /**
     * 是否可以透传点击事件
     */
    fun isClipEventPassThrough(): Boolean {
        return clipHolder.isEventPassThrough()
    }

    /**
     * 是否点击落在裁剪区域
     */
    fun isTouchInClip(x: Float, y: Float): Boolean {
        return clipHolder.getRectF()?.contains(x, y) ?: false
    }

    /**
     * 是否点击落在描述图片区域
     */
    fun isTouchInIntro(x: Float, y: Float): Boolean {
        return descHolder.getRectF()?.contains(x, y) ?: false
    }

    /**
     * 绘制裁剪区或者描述区域
     */
    fun draw(canvas: Canvas, paint: Paint, clip: Boolean, parentWidth: Float, parentHeight: Float) {
        if (clip) {
            clipHolder.draw(canvas, paint, null, parentWidth, parentHeight)
        } else {
            descHolder.draw(canvas, paint, clipHolder.getRectF(), parentWidth, parentHeight)
        }
    }
}